<a href="https://colab.research.google.com/github/Bharat745/H2O/blob/master/Deep_Learning_AutoML_iris.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Deep Learning_AutoML_iris

## Installation and imports


In [0]:
# Installing java version for running H2O on colab
! apt-get install default-jre
!java -version

In [1]:
# Start and connect to local H2O cluster
import h2o
h2o.init(nthreads = -1)

Checking whether there is an H2O instance running at http://localhost:54321 . connected.


--------------------------  ---------------------------------------------------
H2O cluster uptime:         51 mins 26 secs
H2O cluster timezone:       Etc/UTC
H2O data parsing timezone:  UTC
H2O cluster version:        3.24.0.5
H2O cluster version age:    16 days
H2O cluster name:           H2O_from_python_unknownUser_243a22
H2O cluster total nodes:    1
H2O cluster free memory:    3.000 Gb
H2O cluster total cores:    2
H2O cluster allowed cores:  2
H2O cluster status:         locked, healthy
H2O connection url:         http://localhost:54321
H2O connection proxy:
H2O internal security:      False
H2O API Extensions:         Amazon S3, XGBoost, Algos, AutoML, Core V3, Core V4
Python version:             3.6.8 final
--------------------------  ---------------------------------------------------

In [2]:
#Import dataset
url = "http://h2o-public-test-data.s3.amazonaws.com/smalldata/iris/iris_wheader.csv"
iris= h2o.import_file(url)

Parse progress: |█████████████████████████████████████████████████████████| 100%


## Analysis / Modelling

In [0]:
# Splitting the dataset to train and test
train, test = iris.split_frame([0.8])

In [4]:
#looking at the datasets
train.summary()
test.summary

sepal_len sepal_wid petal_len petal_wid class 
 
 
 type real real real real enum 
 mins 4.3 2.2 1.1 0.1 
 mean 5.859836065573771 3.0852459016393428 3.7442622950819673 1.1901639344262296 
 maxs 7.9 4.4 6.9 2.5 
 sigma 0.8272886947575492 0.42788585238240207 1.8030300108120507 0.7781531256462235 
 zeros 0 0 0 0 
 missing 0 0 0 0 0 
 0 4.9 3.0 1.4 0.2 Iris-setosa 
 1 4.7 3.2 1.3 0.2 Iris-setosa 
 2 4.6 3.1 1.5 0.2 Iris-setosa 
 3 5.0 3.6 1.4 0.2 Iris-setosa 
 4 5.4 3.9 1.7 0.4 Iris-setosa 
 5 4.6 3.4 1.4 0.3 Iris-setosa 
 6 5.0 3.4 1.5 0.2 Iris-setosa 
 7 4.9 3.1 1.5 0.1 Iris-setosa 
 8 5.4 3.7 1.5 0.2 Iris-setosa 
 9 4.8 3.4 1.6 0.2 Iris-setosa

sepal_len sepal_wid petal_len petal_wid class 
 
 
 5.1 3.5 1.4 0.2 Iris-setosa 
 4.4 2.9 1.4 0.2 Iris-setosa 
 4.6 3.6 1 0.2 Iris-setosa 
 4.7 3.2 1.6 0.2 Iris-setosa 
 5 3.5 1.3 0.3 Iris-setosa 
 5.1 3.8 1.6 0.2 Iris-setosa 
 4.6 3.2 1.4 0.2 Iris-setosa 
 5.2 2.7 3.9 1.4 Iris-versicolor 
 5 2 3.5 1 Iris-versicolor 
 5.9 3 4.2 1.5 Iris-versicolor

<bound method H2OFrame.summary of >

In [5]:
train.nrows

122

In [6]:
test.nrows

28

### Deep Learning Model

In [0]:
from h2o.estimators.deeplearning import H2ODeepLearningEstimator

In [0]:
# Define the features and target
features = list(test.columns)
target = "class"

In [9]:
# Train the Deep Learning Model
mDL = H2ODeepLearningEstimator(seed= 1234)
mDL.train(x= features, y = target ,training_frame = train)
print(mDL)

deeplearning Model Build progress: |██████████████████████████████████████| 100%
Model Details
H2ODeepLearningEstimator :  Deep Learning
Model Key:  DeepLearning_model_python_1562340797463_2


ModelMetricsMultinomial: deeplearning
** Reported on train data. **

MSE: 0.2122751645958932
RMSE: 0.4607332900886295
LogLoss: 1.0339615996634133
Mean Per-Class Error: 0.2170542635658915
Confusion Matrix: Row labels: Actual class; Column labels: Predicted class



Iris-setosa    Iris-versicolor    Iris-virginica    Error     Rate
-------------  -----------------  ----------------  --------  --------
43             0                  0                 0         0 / 43
0              36                 0                 0         0 / 36
0              28                 15                0.651163  28 / 43
43             64                 15                0.229508  28 / 122

Top-3 Hit Ratios: 


k    hit_ratio
---  -----------
1    0.770492
2    1
3    1

Scoring History: 


timestamp            duration    training_speed    epochs    iterations    samples    training_rmse    training_logloss    training_r2    training_classification_error
--  -------------------  ----------  ----------------  --------  ------------  ---------  ---------------  ------------------  -------------  -------------------------------
    2019-07-05 16:29:57  0.000 sec                     0         0             0          nan              nan                 nan            nan
    2019-07-05 16:29:58  0.726 sec   246 obs/sec       1         1             122        0.835341         4.85917             0.0101063      0.704918
    2019-07-05 16:29:58  1.463 sec   1007 obs/sec      10        10            1220       0.460733         1.03396             0.698865       0.229508

Variable Importances: 


variable    relative_importance    scaled_importance    percentage
----------  ---------------------  -------------------  ------------
petal_wid   1                      1                    0.267398
sepal_len   0.924816               0.924816             0.247294
sepal_wid   0.922061               0.922061             0.246557
petal_len   0.892872               0.892872             0.238752

As we see from the above model that 28/122 were classified wrong, giving a prediction accuracy of 77% on the traning data.

### Results

In [10]:
# Using the Deep Learning model to make predictions
p = mDL.predict(test)

deeplearning prediction progress: |███████████████████████████████████████| 100%


In [11]:
p

predict Iris-setosa Iris-versicolor Iris-virginica 
 
 
 Iris-setosa 0.998851 0.00114853 7.84442e-20 
 Iris-setosa 0.998063 0.00193744 1.88171e-18 
 Iris-setosa 0.999923 7.74369e-05 8.02947e-21 
 Iris-setosa 0.998985 0.0010147 4.57945e-19 
 Iris-setosa 0.999253 0.000746841 1.3287e-19 
 Iris-setosa 0.999761 0.000239409 2.63967e-20 
 Iris-setosa 0.999353 0.000646653 2.41184e-19 
 Iris-versicolor 0.000238537 0.99976 1.92181e-06 
 Iris-versicolor 2.22277e-05 0.999978 4.89992e-08 
 Iris-versicolor 1.62074e-05 0.999982 1.51976e-06

In [12]:
# Checking the performance of the model
mDL.model_performance(test)


ModelMetricsMultinomial: deeplearning
** Reported on test data. **

MSE: 0.12873965177905247
RMSE: 0.3588030821760767
LogLoss: 0.4069171947216555
Mean Per-Class Error: 0.2380952380952381
Confusion Matrix: Row labels: Actual class; Column labels: Predicted class



Iris-setosa    Iris-versicolor    Iris-virginica    Error     Rate
-------------  -----------------  ----------------  --------  ------
7              0                  0                 0         0 / 7
0              14                 0                 0         0 / 14
0              5                  2                 0.714286  5 / 7
7              19                 2                 0.178571  5 / 28

Top-3 Hit Ratios: 


k    hit_ratio
---  -----------
1    0.821429
2    1
3    1

When we test our Deep Learning model, it gives us a higher accuracy of prediction i.e. 82% on our test data.

# Auto ML

In [13]:
# using H2O's Auto ML
from h2o.automl import H2OAutoML
aml = H2OAutoML(max_runtime_secs = 30, seed = 1234)
aml.train(x = features, y = target, training_frame = train)

AutoML progress: |████████████████████████████████████████████████████████| 100%


In [14]:
# Print Leaderboard
print(aml.leaderboard)

model_id mean_per_class_error logloss rmse mse 
 
 
 GBM_grid_1_AutoML_20190705_163058_model_2 0.0340224 0.414 0.18992 0.0360696 
 GLM_grid_1_AutoML_20190705_163058_model_1 0.0417743 0.173156 0.206685 0.0427186 
 GBM_4_AutoML_20190705_163058 0.0417743 0.134912 0.185075 0.0342527 
 GBM_3_AutoML_20190705_163058 0.0417743 0.134912 0.185075 0.0342527 
 GBM_1_AutoML_20190705_163058 0.0417743 0.132244 0.188719 0.0356147 
 GBM_2_AutoML_20190705_163058 0.0417743 0.134912 0.185075 0.0342527 
 XGBoost_grid_1_AutoML_20190705_163058_model_3 0.0432817 0.255115 0.246895 0.0609573 
 XGBoost_grid_1_AutoML_20190705_163058_model_1 0.0432817 0.373907 0.319813 0.10228 
 StackedEnsemble_BestOfFamily_AutoML_20190705_163058 0.0495263 0.220046 0.227193 0.0516168 
 StackedEnsemble_AllModels_AutoML_20190705_163058 0.0510336 0.210908 0.220868 0.0487829

In [15]:
# Using the best model to make predictions
best_model = aml.leader
phat = best_model.predict(test)
phat

gbm prediction progress: |████████████████████████████████████████████████| 100%


predict Iris-setosa Iris-versicolor Iris-virginica 
 
 
 Iris-setosa 1 8.46093e-09 8.43225e-09 
 Iris-setosa 0.999999 9.21576e-07 6.12962e-08 
 Iris-setosa 1 8.5853e-09 8.57263e-09 
 Iris-setosa 1 2.17255e-08 1.38771e-08 
 Iris-setosa 1 8.73555e-09 8.70249e-09 
 Iris-setosa 1 8.56807e-09 8.5637e-09 
 Iris-setosa 1 2.13798e-08 1.36113e-08 
 Iris-versicolor 1.28082e-05 0.999974 1.3374e-05 
 Iris-versicolor 2.22393e-07 0.999998 1.31526e-06 
 Iris-versicolor 5.77618e-07 0.999966 3.38083e-05

In [16]:
# Checking the performance of the best model 
best_model.model_performance(test)


ModelMetricsMultinomial: gbm
** Reported on test data. **

MSE: 0.0788788610341951
RMSE: 0.2808538072275238
LogLoss: 0.6993501134999359
Mean Per-Class Error: 0.047619047619047616
Confusion Matrix: Row labels: Actual class; Column labels: Predicted class



Iris-setosa    Iris-versicolor    Iris-virginica    Error      Rate
-------------  -----------------  ----------------  ---------  ------
7              0                  0                 0          0 / 7
0              12                 2                 0.142857   2 / 14
0              0                  7                 0          0 / 7
7              12                 9                 0.0714286  2 / 28

Top-3 Hit Ratios: 


k    hit_ratio
---  -----------
1    0.928571
2    1
3    1

As we see from the above results, when we run our best model among the list of models on the test data, it gives us even higher accuracy of 92%. oly 2 out of 28 are predicted wrong. |